GVFS/GVFS.Common/GitHubUpgrader.cs

```diff
@@ -0,0 +1,749 @@
+            installationSucceeded = false;
+
+            int exitCode = 0;
+            bool launched = this.TryRunInstallerForAsset(GVFSAssetId, out exitCode, out error);
+            installationSucceeded = exitCode == 0;
+
+            return launched;
+        }
+
+        private bool TryRunInstallerForAsset(string assetId, out int installerExitCode, out string error)
+        {
+            error = null;
+            installerExitCode = 0;
+
+            bool installerIsRun = false;
+            string path;
+            string installerArgs;
+            if (this.TryGetLocalInstallerPath(assetId, out path, out installerArgs))
+            {
+                string logFilePath = GVFSEnlistment.GetNewLogFileName(ProductUpgrader.GetLogDirectoryPath(), Path.GetFileNameWithoutExtension(path));
+                string args = installerArgs + " /Log=" + logFilePath;
+                this.RunInstaller(path, args, out installerExitCode, out error);
+
+                if (installerExitCode != 0 && string.IsNullOrEmpty(error))
+                {
+                    error = assetId + " installer failed. Error log: " + logFilePath;
+                }
+
+                installerIsRun = true;
+            }
+            else
+            {
+                error = "Could not find downloaded installer for " + assetId;
+            }
+
+            return installerIsRun;
+        }
+
+        private void TraceException(Exception exception, string method, string message)
+        {
+            EventMetadata metadata = new EventMetadata();
+            metadata.Add("Method", method);
+            metadata.Add("Exception", exception.ToString());
+            this.tracer.RelatedError(metadata, message, Keywords.Telemetry);
+        }
+
+        private bool TryGetLocalInstallerPath(string assetId, out string path, out string args)
+        {
+            foreach (Asset asset in this.newestRelease.Assets)
+            {
+                if (string.Equals(Path.GetExtension(asset.Name), GVFSPlatform.Instance.Constants.InstallerExtension, StringComparison.OrdinalIgnoreCase))
+                {
+                    path = asset.LocalPath;
+                    if (assetId == GitAssetId && this.IsGitAsset(asset))
+                    {
+                        args = GitInstallerArgs;
+                        return true;
+                    }
+
+                    if (assetId == GVFSAssetId && this.IsGVFSAsset(asset))
+                    {
+                        args = GVFSInstallerArgs;
+                        return true;
+                    }
+                }
+            }
+
+            path = null;
+            args = null;
+            return false;
+        }
+
+        private bool IsGVFSAsset(Asset asset)
+        {
+            return this.AssetInstallerNameCompare(asset, ProductUpgrader.GVFSInstallerFileNamePrefix, ProductUpgrader.VFSForGitInstallerFileNamePrefix);
+        }
+
+        private bool IsGitAsset(Asset asset)

```

Comment:
Will some of this code be moved out when we support `NugetUpgrader`?  As best I can tell not all of this looks specific to GitHub.